### PR TITLE
feat(packaging) Creates a python package to simplify importing the schema in python projects.

### PR DESCRIPTION
**Pull request by fpacifici**
We can simplify the lives of those maintaining python projects that need the mypy type for the event by creating a python package.
This means: 
- it can be imported and kept up to date with the standard tooling already in place instead of building custom logic to download the schema. 
- we can remove the complexity of dealing with jsonschema_typed from the client (which requires having the file in the right place otherwise mypy fails (silently when used through VSCode)).

Then we can decide how to import it, whether to publish this to pypi, or to import the package from github or to import the package via submodule.

### CODE REVIEW

File: py/LICENSE
Comment (untitaker) [inline py/LICENSE]: I think we should use MIT or even public domain otherwise we will not be able to ship with SDKs. Also in the case of BSL we need @BYK to setup his stuff for updating Change date.

The license could potentially also apply to the whole repo

**Review by untitaker — APPROVED** [review]

File: py/LICENSE
Comment (fpacifici) [reply py/LICENSE]: I think you are right, this should not be BSL.

File: py/MANIFEST.in
Comment (fpacifici) [inline py/MANIFEST.in:2]: @untitaker do you remember what the reason was for organizing schemas per source of truth? (relay generated schemas are in the relay subdirectory)? Asking because I wonder whether we should group them the same way into modules in the python package.

Comment (untitaker) [reply py/MANIFEST.in:2]: I thought organizing by source of truth would be most helpful as every team would then own its own namespace (as files are easily overwritten).

Perhaps we want recursive copy + autogenerated `__init__`

File: py/LICENSE
Comment (fpacifici) [reply py/LICENSE]: Actually thinking over this again, I suspect it may be safer to start with BSL as it is more restrictive, then we can open it up if/when we need. Restricting it later would be harder.
(I can add the automation to change the date, I figured out how it works for Snuba).
What do you think ?

File: py/MANIFEST.in
Comment (fpacifici) [reply py/MANIFEST.in:2]: This is going to be an important topic to tackle when we will have more schemas. Whether we organize schemas per owner (which is optimal for writing) or conceptually based on the what each schema represents (which is probably better for consuming). 
Now we have one schema only, and for the foreseeable future we will have only one owner and a handful of schemas. We likely can address this later.

File: py/LICENSE
Comment (untitaker) [reply py/LICENSE]: I can't ship BSL in the SDK in any capacity, it will be a hard blocker. Asking `#team-oss` in Slack.

File: py/setup.py
Comment (mattrobenolt) [inline py/setup.py]: I'd suggest making this less strict here since it's a library. If we decide to bump to say 0.8.1 in sentry core, we gotta bump here as well and tag a new release. I assume we can just leave this without any version qualifier or at least safely a range within semver.

File: py/setup.py
Comment (mattrobenolt) [inline py/setup.py]: I'm not going to be consuming this, but this seems pretty restrictive if we wanna use this elsewhere considering we hardly depend on python 3.8 (only in snuba), and our SDKs would never enforce this or our server for a while.

File: py/setup.py
Comment (mattrobenolt) [inline py/setup.py]: Make sure license is bumped here for whatever yall end up doing.

File: py/setup.py
Comment (mattrobenolt) [inline py/setup.py]: I htink we have `oss@sentry.io` @BYK ?

File: py/setup.py
Comment (BYK) [reply py/setup.py]:
```suggestion
    license="MIT",
```

File: py/setup.py
Comment (BYK) [reply py/setup.py]: Yup, thanks for the heads up @mattrobenolt!

```suggestion
    author_email="oss@sentry.io",
```

**Review by BYK — CHANGES_REQUESTED** [review]
License related things need to change before landing

File: py/LICENSE
Comment (BYK) [reply py/LICENSE]: MIT it is as discussed over at Slack.

File: py/setup.py
Comment (BYK) [inline py/setup.py:40]: Shall we be using this version? How do we even version this?

File: py/setup.py
Comment (BYK) [reply py/setup.py]: Should we also move this out to `requirements.txt` or `pyproject.toml` to be future-proof?

File: py/setup.py
Comment (fpacifici) [reply py/setup.py]: I think I can go down to 3.6 or something around that by importing TypeDict from mypy.experimental as we were doing before 3.8.

File: py/setup.py
Comment (fpacifici) [reply py/setup.py:40]: This is a question that still requires a follow up. Me and Markus did not yet define the details of the release process for the schema. Likely we should generate a version either at every commit (to the schema, which is not supposed to change that often and in a backward compatible way) or something similar. Anyway more often than "monthly" so that the consumers of the schema (as of now snuba and probably the data pipelines) can get up to date quickly.
Do you see any issue with that ?

In practice we are not even going to publish this to pypi anytime soon. I heard from Armin they are already building the automation needed to publish packages to pypi for other reasons, so waiting for that. Which means this version number will not be very relevant up until then and we will depend (from snuba) on the latest commit for the time being.

Comment (BYK) [reply py/setup.py:40]: > Likely we should generate a version either at every commit (to the schema, which is not supposed to change that often and in a backward compatible way) or something similar. Anyway more often than "monthly" so that the consumers of the schema (as of now snuba and probably the data pipelines) can get up to date quickly.
Do you see any issue with that ?

Sounds good to me and also makes me think that we may adopt the CalVer schema that `pytz` uses: https://calver.org/#pytz

File: py/LICENSE
Comment (fpacifici) [reply py/LICENSE]: done

File: py/setup.py
Comment (fpacifici) [reply py/setup.py]: done

File: py/setup.py
Comment (untitaker) [reply py/setup.py:40]: calver is fine but we would need to adopt a much faster release schedule than onprem and dependencies, since a deploy of this package to sentry.io is always blocked on a release, much like the Relay Python package.

Or we do calver but continue to depend on git hashes internally, but who's the consumer of the calver releases then?

**Review by untitaker — APPROVED** [review]

File: py/.gitignore
Comment (BYK) [inline py/.gitignore]: Do we really want nested `.gitignore` files?

File: py/setup.py
Comment (BYK) [inline py/setup.py]: This feels like bad practice to me as it clobbers the check out and file system. Do you think we can find a better directory layout instead?

File: py/setup.py
Comment (BYK) [inline py/setup.py:33]: Same as above. The main problem I have with this is we may not have the file at all and setup/build will just move on without even a warning.

File: py/setup.py
Comment (BYK) [reply py/setup.py:40]: We can do whatever we want with CalVer. We can probably do `YY.MM.MICRO` where micro gets incremented until the month flips over. Or we can do what [PYTZ does](https://calver.org/#pytz): `YYYY.MICRO` and keep incrementing the micro version until the year turns.

File: py/setup.py
Comment (BYK) [inline py/setup.py:51]: No classifiers? https://github.com/getsentry/sentry/blob/7180fb02671bfab8302a3f6918a6ef9cd181e3ed/setup.py#L159-L170

File: py/setup.py
Comment (untitaker) [reply py/setup.py:40]: yeah that sounds good

File: py/setup.py
Comment (untitaker) [reply py/setup.py:33]: I can't really think of a simpler way to get this working so perhaps we just do the copyfile unconditionally adn leave it at that

File: py/setup.py
Comment (fpacifici) [reply py/setup.py:51]: RIght, I forgot to come back to this one.

File: py/.gitignore
Comment (fpacifici) [reply py/.gitignore]: My mistake, sorry, I moved this around and forgot to remove it

File: py/setup.py
Comment (fpacifici) [reply py/setup.py:33]: I don't like it either, but I could not find a way to reference from outside the root of the python package. 
Though you may have a point, in that I am not sure this will work properly (find the file) if we reference the package by github commit. Let me verify that.

File: py/setup.py
Comment (fpacifici) [reply py/setup.py:40]: regarding that, while we are not publishing this to pypi right away. Do you think of any argument (when we get to publish) against publishing new releases potentially quite often? I would like we generate a version at every commit. I doubt we will ever have frequent commits on this schema (at least I would hope). Though it may happen at times.

Comment (BYK) [reply py/setup.py:40]: Strong 👍🏻 for a new package for every new commit as long as we don't publish them to PyPI at this frequency (mostly a courtesy to them, doesn't bother me otherwise 😀 )

File: py/setup.py
Comment (fpacifici) [reply py/setup.py:33]: Unfortunately just doing the copy unconditionally would not work (added a comment about this):
- setup.py runs no matter whether we are building the distribution or importing tthe package. So depending on the phase the source file may or may not be there.
- When we build the distribution from source the source file is there, and by copying the file here, the json file will be in the package produced.
- Importing from source (whether gzipped or not) the source file is generally not there, but, when we built the distribution the file was copied where expected so it is available (but we must not try to copy). It works, I tried tot import the gzip
- Importing from pypi is supposed to work as above
- Importing from github would load the entire repo, (tried, it works) thus the source file is there and running the copy will work.

I think this was the intent for relay as well: https://github.com/getsentry/relay/blob/master/py/setup.py#L22-L27

I think a better solution may be to extend the setuptools sdist command. I tried and that part worked, though somehow the MANIFEST file is not copying all the files and I am not yet sure why. I need more investigation there.

Comment (untitaker) [reply py/setup.py:33]: ok let's leave it like this

**Review by untitaker — APPROVED** [review]

File: py/setup.py
Comment (BYK) [inline py/setup.py]: We can probably add more OSes? No reason to limit this to Linux?

**Review by BYK — APPROVED** [review]

File: py/setup.py
Comment (fpacifici) [reply py/setup.py]: done

**Pull request merged**
